### PR TITLE
Individual part weights

### PR DESCRIPTION
**Pull request by N-Maas**

### CODE REVIEW

**Comment by larsgottesbueren** [comment]
The global rollback for FM has an option to violate the balance constraint by a fixed amount (user parameter). We use that since it gives better solutions to intermediately allow balance violations and then rebalance later on. This would have to be adapted as well!

File: mt-kahypar/io/command_line_options.cpp [hunk 1]
```diff
@@ -106,7 +106,13 @@ namespace mt_kahypar {
              "Summarize results in CSV format")
             ("algorithm-name",
              po::value<std::string>(&context.algorithm_name)->value_name("<std::string>")->default_value("MT-KaHyPar"),
-             "An algorithm name to print into the summarized output (csv or sqlplottools). ");
+             "An algorithm name to print into the summarized output (csv or sqlplottools). ")
+            ("use-individual-part-weights",
```
Comment (kittobi1992) [inline mt-kahypar/io/command_line_options.cpp:110]: I think we do not have to expose this parameter to the user. If --part-weights is specified, then set use_individual_part_weights to true. Should be possible with a lambda function (see e.g. --c-type).

File: mt-kahypar/partition/initial_partitioning/recursive_bisection_initial_partitioner.cpp [hunk 2]
```diff
@@ -332,16 +332,40 @@ namespace mt_kahypar {
...
+        HypernodeWeight perfect_weight_p0 = 0;
+        for ( PartitionID i = 0; i < k0; ++i ) {
+          perfect_weight_p0 += ceil(weight_fraction * context.partition.max_part_weights[i]);
+        }
+        HypernodeWeight perfect_weight_p1 = 0;
+        for ( PartitionID i = k0; i < k; ++i ) {
+          perfect_weight_p1 += ceil(weight_fraction * context.partition.max_part_weights[i]);
+        }
+        const double base = max_part_weights_sum / static_cast<double>(perfect_weight_p0 + perfect_weight_p1);
+        bisection_context.partition.epsilon = total_weight == 0 ? 0 : std::min(0.99, std::max(std::pow(base, 1.0 /
```
Comment (kittobi1992) [inline mt-kahypar/partition/initial_partitioning/recursive_bisection_initial_partitioner.cpp:356]: the adaptive epsilon computation here does not consider the hypergraph weight of the current subhypergraph and the original hypergraph. Is this correct?

Comment (kittobi1992) [reply mt-kahypar/partition/initial_partitioning/recursive_bisection_initial_partitioner.cpp:356]: maybe add a comment ...

File: mt-kahypar/partition/context.cpp [hunk 3]
```diff
@@ -205,19 +211,31 @@ namespace mt_kahypar {
...
-      partition.perfect_balance_part_weights.push_back(
-              partition.perfect_balance_part_weights[0]);
-    }
-    partition.max_part_weights.clear();
-    partition.max_part_weights.push_back((1 + partition.epsilon)
-                                         * partition.perfect_balance_part_weights[0]);
-    for (PartitionID part = 1; part != partition.k; ++part) {
-      partition.max_part_weights.push_back(partition.max_part_weights[0]);
+    if (partition.use_individual_part_weights) {
+      ASSERT(static_cast<size_t>(partition.k) == partition.max_part_weights.size());
```
Comment (kittobi1992) [inline mt-kahypar/partition/context.cpp:215]: is not necessary to also calculate an epsilon here based on the weight_fraction? Some algorithms require that, especially the global rollback algorithm that Lars mentioned.

**Comment by kittobi1992** [comment]
> The global rollback for FM has an option to violate the balance constraint by a fixed amount (user parameter). We use that since it gives better solutions to intermediately allow balance violations and then rebalance later on. This would have to be adapted as well!

the global rollback algorithm is fine as long as we calculate an epsilon based on the max part weights (which is currently missing)

File: mt-kahypar/io/command_line_options.cpp (hunk 1 above)
Comment (N-Maas) [reply mt-kahypar/io/command_line_options.cpp]: Yeah, I just copied the interface from KaHyPar but restricting to --part-weights probably makes more sense.

File: mt-kahypar/partition/initial_partitioning/recursive_bisection_initial_partitioner.cpp (hunk 2 above)
Comment (N-Maas) [reply mt-kahypar/partition/initial_partitioning/recursive_bisection_initial_partitioner.cpp:364]: Well, sort of. I have added a lengthier comment now that hopefully clears it up.

File: mt-kahypar/partition/context.cpp (hunk 3 above)
Comment (N-Maas) [reply mt-kahypar/partition/context.cpp:215]: Whoops :sweat_smile:

**Review by kittobi1992 — APPROVED** [review]

**Pull request merged**